### PR TITLE
Plans: Display control variant when clicking 'View all plans' under a tailored plans grid

### PR DESCRIPTION
**Pull request by oswian**
Related to https://github.com/Automattic/martech/issues/3332

Discussion: p5uIfZ-fIJ-p2#comment-23848

Context: We will soon be running an experiment to test different treatments of the plans grid. For further details, please view the project thread (p5uIfZ-fIN-p2).

## Proposed Changes

When clicking 'View all plans' below a tailored version of the plans grid (eg. a newsletter site), as a user assigned to a treatment variant, override the treatment assignment and display the control variant of the plans grid.

CleanShot 2024-09-17 at 10 10 24@2x

## Why are these changes being made?

To avoid confusing users. This experiment (pbxNRc-45y-p2) does not apply to tailored versions of the plans grid (eg. a newsletter site), so we want to avoid displaying treatment changes when clicking 'View all plans'.

## Testing Instructions

* View the plans grid for a newsletter site with a treatment flag enabled
  * eg. `/plans/:site?flags=simplified-features-grid-b`
* Click 'View all plans' below the plans grid
* Check that the displayed plans grid is not a treatment variant

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Comment by oswian** [comment]
cc @Automattic/martech-decepticons for awareness.

File: client/my-sites/plans-features-main/index.tsx [hunk 1]
```diff
@@ -717,6 +719,23 @@ const PlansFeaturesMain = ( {
 		[ simplifiedFeaturesGridExperimentVariant, translate ]
 	);
 
+	const viewAllPlansButton = (
```
Comment (chriskmnds) [inline client/my-sites/plans-features-main/index.tsx:722]: Component names should be capitalised.

Comment (chriskmnds) [reply client/my-sites/plans-features-main/index.tsx:722]: Let's also rename to `EscapeHatch` as that's the general reference term, the CSS class, and name in conditions.

**Comment by chriskmnds** [comment]
@oswian you can wait for a few hours at least before merging. It's surprising to me that not even a review was requested here, only pinged the team in the aftermath. 

Almost always there are comments/suggestions/request for changes making the rounds in these last couple of PRs I've seen on this functionality/experiment. If that's irritating to the point of starting to ignore the need altogether for code reviews, then I am not sure I can think of better solution than to make code reviews back to a requirement.

cc @southp

File: client/my-sites/plans-features-main/index.tsx (hunk 1 above)
Comment (oswian) [reply client/my-sites/plans-features-main/index.tsx:722]: > Component names should be capitalised.

This isn't a Component definition (which should use capitalisation), but rather a JSX Element.

Comment (chriskmnds) [reply client/my-sites/plans-features-main/index.tsx:722]: It is not indeed. Can we move this out of the `PlansFeaturesMain` component? and would turning it into a component make sense?

File: client/my-sites/plans-features-main/hooks/use-simplified-features-grid-experiment.ts [hunk 2]
```diff
@@ -15,24 +15,31 @@ interface Params {
 
 function useSimplifiedFeaturesGridExperiment( { flowName, isInSignup, intent }: Params ): {
 	isLoading: boolean;
+	isTargetedView: boolean;
```
Comment (chriskmnds) [inline client/my-sites/plans-features-main/hooks/use-simplified-features-grid-experiment.ts:18]: Do you mind explaining what exactly this is referring to? Perhaps adding a comment or considering a more specific name might help.

**Comment by oswian** [comment]
Hi @chriskmnds. I think as a team we should always feel free to review PR's regardless of the merge status. For smaller PR's like this where we feel confident that the changes are low-risk, and where there's some urgency behind launching an experiment, I believe post-merge reviews can often make sense.

I sense some assumptions in your comment -- 'ignore the need altogether for code reviews'. This isn't my approach at all. As you mentioned, the code review process for the [initial PR](https://github.com/Automattic/wp-calypso/pull/94044) involved quite a few comments/suggestions, and I very much appreciate the feedback. I added the team ping on this PR specifically to make sure it was on the team's radar.

Also just for context, I'm fairly new to the concept of merging without a review. In fact joining Dotcom is my first experience with this approach, and it still makes me a little nervous! However I do now see some of the benefits. If as a team we decide that a different approach makes more sense, I think we just need to carefully balance the pro's and con's of each approach.

File: client/my-sites/plans-features-main/hooks/use-simplified-features-grid-experiment.ts (hunk 2 above)
Comment (oswian) [reply client/my-sites/plans-features-main/hooks/use-simplified-features-grid-experiment.ts:18]: Yes happy to follow this up with a comment. It refers to the views that are targeted by this experiment ie. `/start/plans` and `/plans/:site`. I don't think the previous variable name `isEligible` conveyed this as clearly.

**Comment by chriskmnds** [comment]
> Hi @chriskmnds. I think as a team we should always feel free to review PR's regardless of the merge status. For smaller PR's like this where we feel confident that the changes are low-risk, and where there's some urgency behind launching an experiment, I believe post-merge reviews can often make sense.

1. I don't think the changes here are low risk
2. This isn't so urgent as a day in wait would feel blocking
3. Unless you explicitly request a review, then it can easily slip through folks. Requesting a review IS a requirement at the very least.
4. We may well be the only team/division in A8C that brings this mentality of "no approval". and it might be time to reconsider cause we most certainly don't work in low-risk code bases.

> I sense some assumptions in your comment -- 'ignore the need altogether for code reviews'. This isn't my approach at all. As you mentioned, the code review process for the [initial PR](https://github.com/Automattic/wp-calypso/pull/94044) involved quite a few comments/suggestions, and I very much appreciate the feedback. I added the team ping on this PR specifically to make sure it was on the team's radar.

Please request a review in the regular manner before merging anything. and please attempt to allow it some time for folks to see it. post-merge reviews carry a whole different vibe to them

> Also just for context, I'm fairly new to the concept of merging without a review. In fact joining Dotcom is my first experience with this approach, and it still makes me a little nervous! However I do now see some of the benefits. If as a team we decide that a different approach makes more sense, I think we just need to carefully balance the pro's and con's of each approach.

I am a bit confused. If I'm nervous (so "anxious"?) about something, I'd put a few guards around it - like definitely requesting a review, waiting at least a day, communicating in the PR that I've tested thoroughly and I'm about to merge, etc.

File: client/my-sites/plans-features-main/hooks/use-simplified-features-grid-experiment.ts (hunk 2 above)
Comment (chriskmnds) [reply client/my-sites/plans-features-main/hooks/use-simplified-features-grid-experiment.ts:18]: "eligibility" is a common term we've been using ever since I recall, and it immediately "clicks" to me. I wouldn't complicate things by introducing new names if that's what this refers to and nothing else. "target view for experiment" is more ambiguous to me that "eligible for experiment"

File: client/my-sites/plans-features-main/index.tsx [hunk 3]
```diff
@@ -717,6 +719,23 @@ const PlansFeaturesMain = ( {
...
+		<div className="plans-features-main__escape-hatch">
+			<Button
+				borderless
+				onClick={ () => {
+					if ( ! isTargetedViewForSimplifiedFeaturesGridExperiment ) {
+						setSimplifiedFeaturesGridExperimentVariantOverride( 'control' );
+					}
```
Comment (chriskmnds) [inline client/my-sites/plans-features-main/index.tsx:729]: This could have been a side effect on `forceDefaultPlans` changing, no? So we could avoid the changes to the tree nodes and introduce a separate effect that we'd discard when done. Would that be cleaner?

**Comment by oswian** [comment]
> I am a bit confused. If I'm nervous (so "anxious"?) about something, I'd put a few guards around it - like definitely requesting a review, waiting at least a day, communicating in the PR that I've tested thoroughly and I'm about to merge, etc.

@chriskmnds I think a certain level of nervousness/anxiety is normal (and probably a healthy sign) when merging code. It has certainly been the case for me that the more experience I've gained, the more awareness I have for how things can break unexpectedly -- even after a detailed peer review step.

I can certainly double-check to make sure the reviewer auto-assignment occurs (I'm not actually sure why it did not for this PR). However in the spirit of assuming positive intent, rest assured that the [comment](https://github.com/Automattic/wp-calypso/pull/94604#issuecomment-2354464688) I added was intended to achieve the same result.

**Comment by southp** [comment]
> @oswian you can wait for a few hours at least before merging. It's surprising to me that not even a review was requested here, only pinged the team in the aftermath.
> Almost always there are comments/suggestions/request for changes making the rounds in these last couple of PRs I've seen on this functionality/experiment. If that's irritating to the point of starting to ignore the need altogether for code reviews, then I am not sure I can think of better solution but to make code reviews back to a requirement.
> cc @southp

Thanks for looping me in :) 

I feel this is an incident combining three factors:
1. The pressure from the stakeholders to expedite things.
2. The perception that our team would be automatically assigned as reviewers whenever we create a PR -> I don't think we have such thing. The only occasion should be that when the touched code containing the parts that we are set as the code owners 🤔 
3. Being the sole developer on a demanding project like this makes the 1st factor worse.

Before going back to make code reviews required, I think there are still something we can do:

1. Make sure the code review request is made. We can experiment on automatically assigning our team as the reviewer, but that's just a way to make fulfilling this requirement easier. We should always make sure to tag relevant people.
2. Allocate the buffer time for the tagged people to react since we are a distributed team. 
3. This one is for me: even when a project is meant to be mainly worked on by one developer, there should be a 2nd person assigned as the direct support. The direct support is expected to assist on discussion, timely code review, etc. The y can have their own primary initiatives so won't be expected to have major dev contribution. @aneeshd16 once reminded me the importance of this by the diving term: "No one dives alone". This PR is a strong reminder for me. 

cc @Automattic/martech-decepticons here since the above relates to all of us. If there is anything that any of you would like to expand, please feel free to follow up here. I can later summarize everything into a P2 post :)

**Comment by oswian** [comment]
As the newest member of the team I think I might be lacking some context, however I'm still unsure about whether we are practicing optional code reviews. To be clear I don't mind which approach we adopt -- I think there are pros/cons to each, and as the newest team member I'm happy to defer and adopt whichever approach works best for the team.

However I think it's important that we are really clear about how we refer to our approach, because if we refer to it as 'optional' while in reality existing somewhere in-between (ie. with various caveats), I think there is potential for confusion and misinterpretation.

For example, if we require 'buffer time for the tagged people to react' and a direct support who 'is expected to assist on discussion, timely code review', then in my mind it is no longer an optional code review process. It would make sense in that scenario to remove any ambiguity and refer to the review process as mandatory.

Just to further clarify my perspective -- I think there are certainly PR's where reviews should be requested (ideally by multiple people). An example of this is the main [PR](https://github.com/Automattic/wp-calypso/pull/94044) for this experiment. Then there are PR's like this with a much smaller scope and arguably lower risk, which I think a post-merge review and followup is fine. This is a calculation that the PR author takes responsibility for, in terms weighing up any risks with project requirements/timeline etc.

**Comment by southp** [comment]
Thanks for the feedback, Adam :)

I’m sensing some ambiguities stemmed from the nature of code review being optional here. Thus, my intention was to put some conditions in place so it’s clearer when it’s fine to do that. However, I can now see how the current version would add ambiguity rather than clarity, since the detail itself is vague: e.g. by adding buffer time, how much is considered enough? Like you said, it would be better to make it simpler. Also, having 2 people on one project or not is actually another topic. I might still want to pursue that, but it should unrelate to code review and more about better work coverage.

In summary, assigning the team as the reviewer will ensure changes are on the team’s radar and post-merge reviews can occur, so let’s make sure to do that.

As optional code reviews are still a MarTech policy, I trust y’all’s best judgment on what to deploy with/without a review, and will own the responsibility together if there is any repercussion.

**Comment by oswian** [comment]
Thanks @southp, it's great to clarify our process. All sounds good to me!
